ouroboros-consensus/src/ouroboros-consensus/Ouroboros/Consensus/Peras/Voting.hs

```diff
@@ -0,0 +1,196 @@
+  , perasVotingRule1B
+  , perasVotingRule2A
+  , perasVotingRule2B
+  , perasVotingRule1
+  , perasVotingRule2
+  , isPerasVotingAllowed
+  )
+where
+
+import Data.Coerce (coerce)
+import Data.Function (on)
+import Data.Set (Set)
+import GHC.Word (Word64)
+import Ouroboros.Consensus.Block (SlotNo (..))
+import Ouroboros.Consensus.Block.Abstract (Point, StandardHash)
+import Ouroboros.Consensus.Block.SupportsPeras
+  ( PerasRoundNo (..)
+  , ValidatedPerasCert
+  , getPerasCertRound
+  )
+import Ouroboros.Consensus.BlockchainTime.WallClock.Types (WithArrivalTime)
+import Ouroboros.Consensus.Peras.Params
+  ( PerasCertArrivalThreshold (..)
+  , PerasCooldownRounds (..)
+  , PerasIgnoranceRounds (..)
+  , PerasParams (..)
+  )
+import Ouroboros.Consensus.Util.IOLike (MonadSTM (..))
+import Ouroboros.Network.AnchoredFragment (AnchoredFragment)
+
+{-------------------------------------------------------------------------------
+  Voting state
+-------------------------------------------------------------------------------}
+
+-- | Interface needed to compute Peras voting rules
+data PerasVotingView blk = PerasVotingView
+  { pvvPerasParams ::
+      PerasParams
+  -- ^ Peras protocol parameters
+  , pvvAllChains ::
+      Set (AnchoredFragment blk)
+  -- ^ All known chains fragments
+  , pvvLatestCertSeen ::
+      WithArrivalTime (ValidatedPerasCert blk)
+  -- ^ The most recent certificate seen by the voter.
+  , pvvLatestCertOnChain ::
+      WithArrivalTime (ValidatedPerasCert blk)
+  -- ^ The most recent certificate present in some chain.
+  , pvvRound ::
+      SlotNo ->
+      PerasRoundNo
+  -- ^ Get the Peras round number corresponding to a slot.
+  , pvvRoundStart ::
+      SlotNo ->
+      SlotNo
+  -- ^ Get the slot number corresponding to the start of the Peras round for a slot.
+  , pvvBlockSelection ::
+      SlotNo ->
+      Point blk
+  -- ^ The most recent block on the preferred chain that is at least L slots old.
+  , pvvExtends ::
+      Point blk ->
+      WithArrivalTime (ValidatedPerasCert blk) ->
+      Set (AnchoredFragment blk) ->
+      Bool
+  -- ^ Does this block extend the one boosted by a certificate and is present in some chain?
+  , pvvArrivalSlot ::
+      WithArrivalTime (ValidatedPerasCert blk) ->
+      SlotNo
+  -- ^ Transform the arrival time of a certificate into its corresponding slot number
+  }
+
+-- | Run an STM computation that uses a 'PerasVotingView'
+--
+-- | TODO replace some of these inputs with computations derived from the context
+withPerasVotingView ::
+  PerasParams ->
+  Set (AnchoredFragment blk) ->
+  WithArrivalTime (ValidatedPerasCert blk) ->
+  WithArrivalTime (ValidatedPerasCert blk) ->
+  (SlotNo -> PerasRoundNo) ->
+  (SlotNo -> SlotNo) ->
+  (SlotNo -> Point blk) ->
+  (Point blk -> WithArrivalTime (ValidatedPerasCert blk) -> Set (AnchoredFragment blk) -> Bool) ->
+  (WithArrivalTime (ValidatedPerasCert blk) -> SlotNo) ->
+  (PerasVotingView blk -> STM m a) ->
+  STM m a
```

Comment:
This is a sketch for now. The idea is to derive most of these inputs from the context (e.g. a ChainDB instance)